### ISSUE TITLE
Async in MP for "current" Build

### ISSUE THREAD

**Issue by Shawn8901** [Return-To-The-Roots/s25client#757]
Today we got an async error while playing a multiplayer game with 4 humans. 
Ive read the wiki article about asyncs and hope that we got all needed informations.

Im unsure to analyse/fix the problem by myself.

[async_2017-11-17_21-20-17Server.log](https://github.com/Return-To-The-Roots/s25client/files/1483928/async_2017-11-17_21-20-17Server.log)

[async_2017-11-17_21-20-14.zip](https://github.com/Return-To-The-Roots/s25client/files/1483930/async_2017-11-17_21-20-14.zip)

**Comment by Flamefire** [comment]
Thank you! Could you also provide replays from all players please?

**Comment by Shawn8901** [comment]
Find the replays attached.

[replays.zip](https://github.com/Return-To-The-Roots/s25client/files/1485905/replays.zip)

**Comment by Flamefire** [comment]
The current async logs are not that good. I have an improvement in the pipeline. 

Could I get the log of the server? There should be a line `Async at GF xxx of players...`

**Comment by Flamefire** [comment]
I was able to reproduce this by running it in a Linux VM. On Windows 32/64 the problem does not occur. So far there is an extra event in GF 57590 in the Linux version. This would have been detected way earlier in the new version.

**Comment by Flamefire** [comment]
Found and fixed. Thank you for your help!

**Comment by Shawn8901** [comment]
So do i understand right, that you asume in comment above, that the bug should not occur on windows builds? 

im just asking, because all 4 clients are running on a windows build which was build my be from source code aournd the  15th nov with VS2015.  

In the commit you have submitted i dont see any OS specific code, so im just asking.
In general its possbile that this happend, because i know that 3 of the 4 player are using the tool ordering .

If im getting this comment wrong, please ignore mine. :-)

**Comment by Flamefire** [comment]
Asyncs are always hard to debug. So in this case: There are 2 outcomes: one where an extra ware is added to a road causing an extra RANDOM call as seen in the async log. Having the counter# 853997 I found this to be in GF 58864. I tried to reproduce those with VS at the same version (the revision is stored in the replay) but every time that random call did not happen and I had the same behaviour as the "Server" ("S" in the log). Only when I built and executed in on linux I get (again only!) the "C" behaviour where the ware is added.

From here it was easy: I printed the number of events in every GF which led me to GF 57589, the last GF where the event numbers match (used a diff tool to compare the huge resulting logs). So then I printed all events added in GF 57589. This showed an extra event in the linux version. That event is even added AFTER GF 57589 but before 57590 is executed. So jump to 57589 and breakpoint on the addEvent function. Backtrace showed a metalworker that starts producing in reaction to a tool order. Single-stepping the tool order in windows and linux showed, that in linux the metalworker is notified but in windows he isn't. Debugging the boost.Lambda used is a mess but good guessing led to `metalworksPlayer != toolOrderPlayer` although they should have been equal. Looking at the notification code solved the mystery: The ctor of `nofMetalWorker` receives the owner in argument `player` and stores this in the instance variable `player` which works in the ctor initializer list although they have the same name. However the lambda uses a reference to `player` (in case the figure would change its owner which is currently impossible, but who knows...) but of course this is a reference to the argument. After the function its value is undefined and can be anything.
So what happened: In the linux version compiler/runtime/moon phase decided that the variable at that point has its original value again. The windows version had something different. Hence the different outcome.

Simple fix: use `this->player` to use the instance variable.

Final note: Variable shadowing bites you in the ass at some point. But unless we prefix or suffix the instance members I see no good solution for ctors.
